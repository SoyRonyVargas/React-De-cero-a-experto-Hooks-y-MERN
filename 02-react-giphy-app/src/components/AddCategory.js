import React, { useState } from 'react'
import PropTypes from 'prop-types';

export const AddCategory = ({ setCategories }) => {

    const [inputValue, setInputValue] = useState(''); // ''

    const handleInputChange = ( e ) => {
        
        console.log('handle llamado');

        setInputValue( e.target.value );

    }

    const handleSubmit = (e) => {
        
        e.preventDefault();

        console.log('formulario prevenido' , inputValue);

        if ( inputValue.trim().length > 2 ) {
            setCategories( cats => [ inputValue, ...cats, ] );
            setInputValue('');
        }

    }

    return (
        <form onSubmit={ handleSubmit }>
            <p id="valor"> { inputValue } </p>
            <input 
                type="text"
                value={ inputValue }
                onChange={ handleInputChange }
            />
        </form>
    )
}


AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
