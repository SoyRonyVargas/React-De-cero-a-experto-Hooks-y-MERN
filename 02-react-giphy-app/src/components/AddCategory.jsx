import React, { useState } from 'react'
import PropTypes from 'prop-types';

export const AddCategory = ({ setCategories }) => {

    const [inputValue, setInputValue] = useState('');

    const handleInputChange = ( e ) => {
        
        setInputValue( e.target.value );

    }

    const handleSubmit = (e) => {
        
        e.preventDefault();

        if ( inputValue.trim().length > 2 ) {
            setCategories( cats => [ inputValue, ...cats, ] );
            setInputValue('');
        }

    }

    return (
        <form aria-label='form' onSubmit={ handleSubmit }>
            <p id="valor"> { inputValue } </p>
            <input 
                type="text"
                value={ inputValue }
                onChange={ handleInputChange }
            />
        </form>
    )
}


AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
