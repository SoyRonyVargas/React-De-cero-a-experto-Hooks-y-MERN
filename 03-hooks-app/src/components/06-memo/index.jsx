import { useState } from "react";
import useCounter from "../../hooks/useCounter"
import Small from './Small'

const MultipleHooks = () => {
  
    const { counter, increment } = useCounter(1);
    const [ show, setShow ] = useState(false);

    return (
        <div className="">
        <h1>Memo</h1>
        <Small value={counter} />
        <hr />
        
        <button onClick={increment} className="btn btn-primary mx-1">
            Next 
        </button>
        <button onClick={() => {
            setShow(!show)
        }} className="btn btn-primary">
            Show / Hide { JSON.stringify(show , null , 3)} 
        </button>
        </div>
    );
};

export default MultipleHooks;
