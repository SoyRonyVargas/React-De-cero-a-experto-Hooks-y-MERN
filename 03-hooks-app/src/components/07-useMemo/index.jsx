import { useMemo, useState } from "react";
import useCounter from "../../hooks/useCounter"

const MultipleHooks = () => {
  
    const { counter, increment } = useCounter(1);
    const [ show, setShow ] = useState(false);

    const process = ( it = 1 ) => {

        for( let i = 0 ; i < it ; i++ )
        {
            console.log('iteracion...');
        }

        return it + ' iteraciones realizadas.'

    }

    const memoProcess = useMemo( () => process(counter) , [ counter ] )	

    return (
        <div className="">
            
            <h1>useMemo {counter} </h1>
            
            <hr />

            <p>
                { memoProcess }
            </p>
            
            <button onClick={increment} className="btn btn-primary mx-1">
                Next 
            </button>
            <button onClick={() => {
                setShow(!show)
            }} className="btn btn-primary">
                Show / Hide { JSON.stringify(show , null , 3)} 
            </button>
        </div>
    );
};

export default MultipleHooks;
