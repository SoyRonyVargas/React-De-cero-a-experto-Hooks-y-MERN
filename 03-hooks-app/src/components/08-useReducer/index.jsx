import React, { useReducer , useRef } from 'react'
import TodoList from './TodoList'
import TodoAdd from './TodoAdd'
import reducer from './reducer'

const index = () => {

    const ref = useRef()

    const initialState = [
        {
            id: new Date().getTime(),
            desc: 'Aprender React',
            done: false
        }
    ]

    const [ todos , dispatch ] = useReducer(reducer, initialState)

    const handleSubmit = (e) => {

        e.preventDefault()

        const newTodo = {
            id: new Date().getTime(),
            desc: ref.current.value,
            done: false
        }

        ref.current.value = ''

        const action = {
            type: 'ADD_TODO',
            payload: newTodo
        }

        dispatch(action)

    }

    const handleDelete = ( id ) => {

        const action = {
            type: 'delete',
            payload: id
        }

        dispatch(action)

    }

    return (
        <div>
            <h1> Todos - {todos.length} </h1>
            <hr />
            <main className='row '>
                
                <TodoList
                    handleDelete={handleDelete}
                    todos={todos}
                />

                <TodoAdd
                    handleSubmit={handleSubmit}
                    ref={ref}
                />

            </main>
        </div>
    )

}

export default index