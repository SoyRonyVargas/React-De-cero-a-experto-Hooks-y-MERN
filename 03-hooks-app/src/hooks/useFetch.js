import { useEffect, useState } from 'react'

const useFetch = (url) => {
    
    const [ state , setState ] = useState({
        data: null,
        loading: true,
        error: null
    })

    useEffect(() => {
        
        getData()

        return () => {

        }

    }, [url])

    const getData = async () => {

        setState({
            ...state,
            loading: true
        })

        const response = await fetch(url)

        const data = await response.json()

        setTimeout( () => {
            setState({
                ...state,
                data: data,
                loading: false
            })
        }, 0)

    }

    return state
}

export default useFetch
