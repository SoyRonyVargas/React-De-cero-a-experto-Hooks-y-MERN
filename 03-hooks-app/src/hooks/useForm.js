import { useState , useEffect } from 'react'

const useForm = () => {
    
    const [ values, setValues ] = useState({
        name: '',
        email: '',
        password: ''
    })

    useEffect( () => {

        console.log('email cambio');

    } , [values.email])

    const handleSubmit = (e) => {

        e.preventDefault()
        
        console.log(values);

    }

    const handleInputChange = ( event ) => {

        const { name , value } = event.target

        setValues({
            ...values,
            [name]: value
        })

    }


    return [ values , handleInputChange , handleSubmit ]

}

export default useForm
