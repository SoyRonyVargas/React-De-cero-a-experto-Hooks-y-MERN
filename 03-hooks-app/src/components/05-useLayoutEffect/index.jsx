import useCounter from "../../hooks/useCounter";
import useFetch from "../../hooks/useFetch";
import React , { useLayoutEffect , useRef } from "react";

const MultipleHooks = () => {
  const { counter, increment } = useCounter(1);

  const { data, loading } = useFetch(
    `https://www.breakingbadapi.com/api/quotes/${counter}`
  );

  const { quote } = !!data && data[0];

  const ref = useRef();

  useLayoutEffect( () => {

    console.log(ref.current.getBoundingClientRect());

  }, [quote])

  return (
    <div className="">
      <h1>useLayoutEffect</h1>
      <hr />
      <figure className="card p-3">
        <blockquote className="blockquote">
          <p ref={ref}> {quote} </p>
        </blockquote>
      </figure>
      <button onClick={increment} className="btn btn-primary">
        Next Quote
      </button>
    </div>
  );
};

export default MultipleHooks;
