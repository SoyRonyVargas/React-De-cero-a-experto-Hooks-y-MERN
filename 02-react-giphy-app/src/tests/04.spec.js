import { render , screen } from '@testing-library/react'
import { GifGrid } from '../components/GifGrid'
import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs'
import '@testing-library/jest-dom'

jest.mock('../hooks/useFetchGifs')

describe('Probando el componente GifGrid', () => {

    const data = [
       {
           id: 1,
           title: 'Hello',
           url: 'https://www.udemy.com/staticx/udemy/images/v7/logo-udemy-inverted.svg'
       }, 
       {
           id: 2,
           title: 'Hello mario',
           url: 'https://www.udemy.com/staticx/udemy/images/v7/logo-udemy-inverted.svg'
       }, 
       {
           id: 3,
           title: 'Hello mario',
           url: 'https://www.udemy.com/staticx/udemy/images/v7/logo-udemy-inverted.svg'
       } 
    ]

    test('Deberia generar el snapshot', () => {
        
        useFetchGifs.mockImplementation( () => {
            return {
                data: [],
                loading: true
            }
        }) 
        
        const { container } = render(<GifGrid category={'goku'} />)

        expect(container).toMatchSnapshot()

    })

    test('Debe de mostrar items cuando se cargan imagenes', () => {
        
        useFetchGifs.mockImplementation( () => {
            return {
                data,
                loading: false
            }
        }) 

        const { container } = render(<GifGrid category={'goku'} />)

        // BUSCAMOS QUE EL PARRAFO DE LOADING NO ESTE MONTADO

        const loader = container.querySelector("animate__flash")

        const tot = screen.getAllByRole("img").length

        expect( loader ).toBe( null )

        expect( tot ).toBe(3)
        
    })
    
})
