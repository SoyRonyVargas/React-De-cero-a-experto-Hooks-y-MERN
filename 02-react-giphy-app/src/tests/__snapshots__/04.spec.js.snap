// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probando el componente GifGrid Deberia generar el snapshot 1`] = `
<div>
  <h3
    class="animate__animated animate__fadeIn"
  >
     
    goku
     
  </h3>
  <p
    class="animate__animated animate__flash"
  >
    Loading
  </p>
  <div
    class="card-grid"
  />
</div>
`;
